appcues/src/main/java/com/appcues/trait/appcues/BackdropKeyholeTrait.kt

```diff
@@ -0,0 +1,145 @@
+import androidx.compose.animation.core.TweenSpec
+import androidx.compose.animation.core.animateFloatAsState
+import androidx.compose.animation.core.tween
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.BoxScope
+import androidx.compose.foundation.layout.fillMaxSize
+import androidx.compose.foundation.shape.RoundedCornerShape
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.remember
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.drawWithContent
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.geometry.Size
+import androidx.compose.ui.graphics.ClipOp
+import androidx.compose.ui.graphics.Path
+import androidx.compose.ui.graphics.addOutline
+import androidx.compose.ui.graphics.drawscope.clipPath
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.dp
+import com.appcues.data.model.AppcuesConfigMap
+import com.appcues.data.model.getConfig
+import com.appcues.data.model.getConfigInt
+import com.appcues.trait.BackdropDecoratingTrait
+import com.appcues.trait.appcues.BackdropKeyholeTrait.ConfigShape.CIRCLE
+import com.appcues.trait.appcues.BackdropKeyholeTrait.ConfigShape.RECTANGLE
+import com.appcues.trait.appcues.StepAnimationTrait.StepAnimationEasing
+import com.appcues.trait.appcues.StepAnimationTrait.StepAnimationEasing.EASE_IN
+import com.appcues.trait.appcues.StepAnimationTrait.StepAnimationEasing.EASE_IN_OUT
+import com.appcues.trait.appcues.StepAnimationTrait.StepAnimationEasing.EASE_OUT
+import com.appcues.trait.appcues.StepAnimationTrait.StepAnimationEasing.LINEAR
+import com.appcues.trait.appcues.TargetElementTrait.TargetRect
+import com.appcues.ui.composables.LocalAppcuesStepMetadata
+import kotlin.math.pow
+import kotlin.math.sqrt
+
+internal class BackdropKeyholeTrait(
+    override val config: AppcuesConfigMap,
+) : BackdropDecoratingTrait {
+
+    companion object {
+
+        const val TYPE = "@appcues/backdrop-keyhole"
+
+        private const val DEFAULT_ANIMATION = 300
+        private const val CORNER_RADIUS_CIRCLE_PIECES = 8
+    }
+
+    private enum class ConfigShape {
+        RECTANGLE, CIRCLE
+    }
+
+    private val emptyRect = TargetRect(0, 0, 0, 0)
+
+    private val shape = when (config.getConfig<String>("shape")) {
+        "circle" -> CIRCLE
+        else -> RECTANGLE
+    }
+    private val rectCornerRadius = config.getConfigInt("cornerRadius") ?: 0
+
+    @Composable
+    override fun BoxScope.BackdropDecorate(content: @Composable BoxScope.() -> Unit) {
+        val density = LocalDensity.current
+        val metadata = LocalAppcuesStepMetadata.current
+
+        val actualRect = remember(metadata) {
+            (metadata.actual[TargetElementTrait.METADATA_TARGET_RECT] as TargetRect?) ?: emptyRect
+        }
+
+        val duration = remember(metadata) {
+            (metadata.actual[StepAnimationTrait.METADATA_ANIMATION_DURATION] as Int?) ?: DEFAULT_ANIMATION
```

Comment:
if no animation trait is present, should we have no animation or some default animation? cc @mmaatttt to keep in sync

Reply:
going for 350millis as the default but open for suggestions